packages/workflow/src/WorkflowDataProxy.ts

```diff
@@ -943,6 +943,13 @@ export class WorkflowDataProxy {
 					throw createExpressionError(`"${nodeName}" node doesn't exist`);
 				}
 
+				if (!that?.runExecutionData?.resultData?.runData.hasOwnProperty(nodeName)) {
+					if (that.workflow.getNode(nodeName)) {
+						throw createExpressionError(`no data, execute "${nodeName}" node first`);
+					}
+					throw createExpressionError(`"${nodeName}" node doesn't exist`);
+				}
+
```

Comment:
Bear in mind this is a breaking change.

Reply:
should move this change into the v1 branch instead?

Reply:
That or we could do what @janober did for the `$node['nodeName']` and make it throw but not stop the execution and leave a note to make it stop the execution for the V1 release. Then, it would not be a breaking change but the exact behavior we have for `$node['nodeName']`   See [here](https://github.com/n8n-io/n8n/blob/3f533873e7516ce636f211e534fcb82ba414e320/packages/workflow/src/WorkflowDataProxy.ts#L595). 

<img width="751" alt="image" src="https://github.com/n8n-io/n8n/assets/16496553/0e11bcc5-ee1e-464f-a61a-9e0ac4dd918e">

Reply:
Also, currently, we have an inconsistency (not sure if expected?) when retrieving a node that does not exist. One throws an error (making the execution fail), and the other does not.

Old syntax: 

<img width="876" alt="Pasted Graphic 28" src="https://github.com/n8n-io/n8n/assets/16496553/7c3a165e-442c-445c-8eb8-8b570f1039e0">

New syntax: 

<img width="971" alt="Pasted Graphic 27" src="https://github.com/n8n-io/n8n/assets/16496553/89d6afb9-38d3-426c-94b4-95a2de0ad183">